src/realm/db.cpp

```diff
@@ -1498,6 +1498,26 @@ bool DB::compact(bool bump_version_number, util::Optional<const char*> output_en
     return true;
 }
 
+void DB::write_copy(StringData path, util::Optional<const char*> output_encryption_key, bool allow_overwrite)
+{
+    // group::write() will throw if the file already exists.
+    // To prevent this, we have to remove the file (should it exist)
+    // before calling group::write().
+    if (allow_overwrite)
+        File::try_remove(path);
+
+    SharedInfo* info = m_file_map.get_addr();
+    const char* write_key = bool(output_encryption_key) ? *output_encryption_key : m_key;
+
+    auto tr = start_write();
+    if (!tr->get_history()->is_clean(tr->get_version())) {
+        throw std::runtime_error("Client changes not integrated in server");
```

Comment:
Should we crash the app if this is the case or just return false?

Reply:
Perhaps let it be up to the SDK if they want to pass the error along or outright crash.

Reply:
Crashing is probably fine, since this code is not a part of normal app flow, and it is likely a bug in the public-facing API if this situation is reached.

The API that developers are using to generate these files should ensure this as an invariant (i.e., not provide any method to make modifications to the local file after downloading from the server).

I would change the exception message to be more descriptive, though. Something like: "The file could not be prepared for bundling because of local changes in the file that do not exist on the server".

Reply:
SDKs can try/catch if they want to handle this otherwise.

Reply:
If we want to support try/catch, we definitely need a different exception type than `std::runtime_error`. :-)

Reply:
Perhaps get input from the users of the API?

Reply:
Rephrased